### ISSUE TITLE
Repo: use `stableTypeOrdering` compiler option

### ISSUE THREAD

**Issue by StyleShit** [typescript-eslint/typescript-eslint#12356]
### Suggestion

Coming from https://github.com/typescript-eslint/typescript-eslint/pull/12139#discussion_r3254110184, I think we could maybe set `stableTypeOrdering: true` in our tsconfig

It should fix some weird union ordering bugs (see above PR comment), and help us get ready for TS 7 (for when that will happen...) 

References:
- https://www.typescriptlang.org/docs/handbook/release-notes/typescript-6-0.html#the---stabletypeordering-flag
- https://github.com/microsoft/TypeScript/pull/63084
- https://x.com/robpalmer2/status/2021684427890258020

_Note:_ Both [the docs](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-6-0.html#the---stabletypeordering-flag:~:text=it%20can%20add%20a%20substantial%20slowdown%20to%20type%2Dchecking%20%28up%20to%2025%25%20depending%20on%20codebase%29%2E) and [the PR](https://github.com/microsoft/TypeScript/pull/63084#issuecomment-3864762468) mention that this flag slows down typechecking, but my benchmarks show otherwise (~25% faster), which could also be a nice addition to #11204 

_Note 2:_ This is not the default for TS 6 users. I'm not sure whether it falls under https://github.com/typescript-eslint/typescript-eslint/pull/12139#pullrequestreview-4304506375 or https://github.com/typescript-eslint/typescript-eslint/pull/12139#issuecomment-4478704221, but worth mentioning

### Additional Info

```
$ hyperfine --prepare "pnpm clean" "pnpm typecheck" "STABLE_TYPE_ORDERING=true pnpm typecheck"

Benchmark 1: pnpm typecheck
  Time (mean ± σ):     15.856 s ± 0.883 s    [User: 27.862 s, System: 3.761 s]
  Range (min … max):   14.644 s … 17.084 s    10 runs

Benchmark 2: STABLE_TYPE_ORDERING=true pnpm typecheck
  Time (mean ± σ):     12.669 s ± 0.763 s    [User: 29.083 s, System: 3.852 s]
  Range (min … max):   12.006 s … 14.469 s    10 runs

Summary
  'STABLE_TYPE_ORDERING=true pnpm typecheck' ran
    1.25 ± 0.10 times faster than 'pnpm typecheck'
```

**Comment by JoshuaKGoldberg** [comment]
I'm in favor. It's going to be the new default for TypeScript folks soon, so making sure we work well with it is important.

**Comment by kirkwaiblinger** [comment]
This part makes me a bit confused whether it's recommended for us? 

> **Note that this flag is only intended to help diagnose differences between 6.0 and 7.0 - it is not intended to be used as a long-term feature**

https://www.typescriptlang.org/docs/handbook/release-notes/typescript-6-0.html#the---stabletypeordering-flag

🤷‍♂️ 

But otherwise seems reasonable

**Comment by StyleShit** [comment]
Yeah, this is quite confusing
I've read it as "this should help you ensure you're compatible with TS 7", but maybe that's wrong

Also, I assumed that the "_not intended to be used as a long-term feature_" is due to the performance degradation (which is not the case for us, but we'll have to quadruple-check it again, just to be sure):
> Note that we don’t necessarily encourage using this flag all the time as it can add a substantial slowdown to type-checking (up to 25% depending on codebase).

In this case, "don’t necessarily" implies that you _can_ do it, but it may impact performance, no?

Should we ping someone from the TS team to shed some light here?

**Comment by JoshuaKGoldberg** [comment]
We're kind of a less-common case: a cross-TS-version library project rather than app code.

> someone from the TS team

I'd maybe post in the TS Discord?

**Comment by StyleShit** [comment]
Got a reply from the TS team (Jake Bailey):

> It's "okay", but I would be pretty surprised if you saw no change at all
>
> -- _[Discord](https://discord.com/channels/508357248330760243/1510916200354414693/1511046439633948673)_

Weirdly enough, I did get ~14% slower runs when running tsc directly in eslint-plugin, but still got the same uplift mentioned in OP when running the whole suite through nx
